Прохожу спринт 11

EntityVadim opened the pull request:
Ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
Устрани замечание ревью: "Чтобы во всем методе не дублировать DispatchQueue.main.async {} было бы круто реализовать расширение, которое completion-ы сразу оборачивает в асинхронный вызов на главном потоке"

EntityVadim committed:
Добавил пакет (https://github.com/relatedcode/ProgressHUD), Использовал ProgressHUD в коде для показать и скрытия индикатор загрузки

EntityVadim committed:
Сделал пункт:
Чтобы избежать ситуации гонки в функции fetchOAuthToken, нужно:
Проверить, выполняется ли в данный момент запрос в сеть (POST-запрос за Auth Token'ом).
Если выполняется, возможны два варианта:
если Auth Code в запросе совпадает со значением code, нужно отменить предыдущий запрос и сделать новый;
если не совпадает, не нужно ничего делать; выйдите из функции, не создавая новых запросов.

EntityVadim committed:
Доработка по прошлому комиту

EntityVadim committed:
Выполнил задание:
Добавьте обработку для потенциальной гонки в ImageFeed по уроку.
Получившееся решение должно корректно отрабатываться в ситуациях, когда:
а) fetchOAuthToken() был вызван два раза подряд с одинаковым code (второй вызов не дожидается завершения первого);
b) fetchOAuthToken() был вызван сначала с одним code, но спустя некоторое время, когда первый запрос ещё не успел завершиться, был вызван с другим значением кода;
c) fetchOAuthToken() всегда возвращает через замыкание результат своей работы: неважно выполнили мы запрос токена или нет, это поможет лучше понимать, как именно отработала функция в точке вызова.

EntityVadim committed:
Выполнил самостоятельную задачу:
Что нужно сделать:
Добавьте обработку для потенциальной гонки, описанной в этом уроке, в ImageFeed.
Затем нужно добавить новый класс UIBlockingProgressHUD, который синхронизирует показ и скрытие ProgressHUD с блокировкой UI.
Сразу после закрытия WebView при авторизации (при получении CODE) нужно показать индикатор активности, используя UIBlockingProgressHUD.
При завершении запроса, независимо от результата, нужно скрыть индикатор активности.
Убедиться, что пользователь не может повторно открыть экран с WebView, пока индикатор активности отображается на экране.

EntityVadim committed:
Делаю Задачу 1.а. Получить базовую информацию о профиле (Все пункты, что прописал, доделаны)
Откройте Xcode проект с ImageFeed.
По аналогии с OAuth2Service:
- Создайте файл ProfileService.swift.
- Добавьте в него класс final class ProfileService.
- Добавьте структуру struct ProfileResult: Codable для декодирования ответа. Какие поля добавить, вам нужно понять самостоятельно, анализируя документацию.
- Добавьте структуру struct Profile, которую мы будем использовать в UI-слое (например, на экране профиля), со следующими полями:
username — логин пользователя в том же виде, в каком мы получаем его от сервиса,
name — конкатенация имени и фамилии пользователя (если first_name = "Ivan", last_name = "Ivanov", то name = "Ivan Ivanov"),
loginName — username со знаком @ перед первым символом (если username = "ivanivanov", то loginName = "@ivanivanov"),
bio совпадает с bio из ProfileResult.
- Опишите вспомогательный метод, который будет создавать URLRequest.

EntityVadim committed:
Доделал Задача 1.а. Получить базовую информацию о профиле:
- В ProfileService добавьте и имплементируйте метод func fetchProfile(_ token: String, completion: @escaping (Result<Profile, Error>) -> Void).
- Метод fetchProfile не должен приводить к гонкам, если он вызывается несколько раз подряд (каждый следующий вызов не дожидается завершения предыдущего).
- На экране ProfileViewController в методе viewDidLoad нужно вызвать метод fetchProfile и обновить лейблы:
nameLabel (например, "Ivan Ivanov"),
loginNameLabel (например, "@ivanivanov"),
и descriptionLabel (совпадает с bio).
- Добавьте вывод сообщений об ошибках в консоль.

EntityVadim committed:
- Сделал рефакторинг кода;
- Выполнил: Задача 1.б. Показать Splash Screen, пока идёт загрузка данных профиля
   - Мы хотим перенести вызов метода fetchProfile из ProfileViewController в SplashViewController. Сразу после получения токена, мы
     сделаем запрос fetchProfile и только после его завершения, перейдём на экран с лентой фотографий, из которой пользователь
     может перейти в свой профиль.
   - Откройте Xcode проект с ImageFeed.
   - Перейдите в файл SplashViewController.swift.
   - Добавьте свойство для хранения ссылки на ProfileService, новый метод fetchProfile(_ token: String), и вызовите его в методе
     didAuthenticate, сразу за закрытием экрана AuthViewController.
- Изменил способ хранения токена

EntityVadim committed:
- Использовал результат fetchProfile в ProfileViewController
- Применил использование синглтона в SplashViewController и ProfileService

EntityVadim committed:
- Организовал код
- Исправил проблему (Не получал данные с сервера)

EntityVadim committed:
- Выношу постепенно обработку ошибок в отдельный файл

EntityVadim committed:
- Перевел ошибки на русский язык

EntityVadim committed:
- Сделал Задача 2.a. Получить аватарку пользователя
  - По аналогии с ProfileService:
  - Создайте файл ProfileImageService.swift.
  - Добавьте в него класс final class ProfileImageService.
  - Для декодирования ответа добавьте структуру struct UserResult: Codable (из JSON объекта. Нам нужно получить только
    содержимое profile_image и достать из него URL для маленькой картинки — small.
  - В ProfileImageService добавьте и имплементируйте метод  func fetchProfileImageURL(username: String, _ completion: @escaping
    (Result<String, Error>) -> Void) для получения URL для маленькой версии аватарки пользователя.
  - Добавьте в ProfileImageService синглтон — статическое свойство static let shared = ProfileImageService() и приватный
    инициализатор.
  - В ProfileImageService запишите свойство private (set) var avatarURL: String?, куда подставляется profile_image в случае
    успешного запроса.
  - Метод fetchProfileImageURL не должен приводить к гонкам в случае, если он вызывается несколько раз подряд (каждый
    следующий вызов не дожидается завершения предыдущего).
- Задача 2.б. Вызвать метод fetchProfileImageURL
  - После получения username в методе fetchProfile, добавьте вызов метода fetchProfileImageURL в класс SplashViewController,
    используя ProfileImageService.shared

EntityVadim committed:
- Добавил пакет Kingfisher
- Добавил новое имя нотификации, в методе fetchProfileImageURL сразу после вызова completion добавил публикацию
  нотификации
- Добавил обсервер в методе viewDidLoad

EntityVadim committed:
- Выполнил задачи:
Задача 1. Заменить «старое» API для KVO на «новое» в WebViewViewController
Пройдите в класс WebViewViewController и замените «старое» API для KVO (WKWebView.estimatedProgress) на «новое».
Задача 2. Сетевой запрос (на дженериках)
Что сделать:
Напишите имплементацию для метода objectTask(for:) так, чтобы он создавал задачу на запрос с помощью data(for:).
Если запрос в  data(for:) вернет ошибку, «прокиньте» её в completion метода objectTask(for:).
Задача 3. Заменить использование data(for:) на objectTask(for:) (ЧАСТИЧНО СДЕЛАНО)
Замените использование data(for:) на objectTask(for:) в сервисах:
OAuth2Service,
ProfileService,
ProfileImageService.

EntityVadim committed:
- Выполнил задачу:
Задача 5. Показать алерт с ошибкой в AuthViewController
При получении ошибки в сетевом запросе в AuthViewController нужно показать UIAlertController с заголовком «Что-то пошло не так(», подписью «Не удалось войти в систему» и кнопкой «Ок» для скрытия алерта.

EntityVadim committed:
- Пытаюсь выполнить задачу
Задача 3. Заменить использование data(for:) на objectTask(for:)
Замените использование data(for:) на objectTask(for:) в сервисах:
OAuth2Service,
ProfileService,
ProfileImageService.

EntityVadim committed:
- Объеденил два файла Profile и ProfileResult

EntityVadim committed:
- Исправил мелкие замечания

EntityVadim committed:
- Пытаюсь выполнить самостоятельную задачу:
  - Что нужно сделать:
    1. Подключить библиотеку Kingfisher к нашему приложению с помощью SPM.
    2. Используйте Kingfisher для загрузки аватарки профиля.

EntityVadim committed:
- Сделал задание по прошлому комиту

EntityVadim committed:
Удалил файл

EntityVadim committed:
- Увеличил обработку ошибок
- Избавился от fatalError

EntityVadim committed:
Исправления по прошлому коммиту

EntityVadim committed:
- Создал View Controller из кода (TabBarController)
- Сделал аватарку круглой, вместо квадратной

EntityVadim committed:
- Вызвал switchToTabBarController и fetchProfile в главном потоке
- Настроил UIWindow в AppDelegate иSceneDelegate

EntityVadim committed:
- Сделал самостоятельную задачу:
  - Сейчас у нас только ProfileViewController свёрстан в коде. В качестве домашнего задания предлагаем сверстать в коде
    SplashViewController, создавая его также из кода.
    - Для этого нужно будет:
    - Удалить SplashViewController из Storyboard.
    - В файле SplashViewController.swift сверстать в коде содержимое контроллера. А именно — добавить UIImageView на View,
      выставить ему image равный splash_screen_logo и констрейнтами разместить по центру экрана.
    - В SceneDelegate.swift изменить создание контроллера на SplashViewController().
    - Когда мы удалим контроллер из Storyboard, то Segue для перехода на экран авторизации также удалится. Значит, нам нужно
      поменять навигацию: добавить Storyboard ID для AuthViewController, а из SplashViewController.swift удалить метод prepare(for
      segue: UIStoryboardSegue, sender: Any?).
    - Вместо performSegue(withIdentifier: showAuthenticationScreenSegueIdentifier, sender: nil) нужно:
         - создать AuthViewController.
         - выставить делегатом себя viewController.delegate = self.
         - выставить ему modalPresentationStyle равный .fullScreen;
презентовать его, используя present(_:animated:completion:).

EntityVadim committed:
- Исправил ошибку (Failed to instantiate the default view controller for UIMainStoryboardFile 'Main' - perhaps the designated entry point is not set?)

EntityVadim committed:
- Пытаюсь исправить черный экран

EntityVadim committed:
- Исправил черный экран

EntityVadim committed:
- Исправил проблемы (вечная загрузка, ругался на получения окна, реализовал метод выхода на окно авторизации)

EntityVadim committed:
- Изменения по прошлому коммиту

EntityVadim committed:
- Решил проблему с двойным получением токена

EntityVadim committed:
- Пытаюсь реализовать кнопку выход

helecta commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -62,13 +70,18 @@ extension AuthViewController: WebViewViewControllerDelegate {
-        delegate?.authViewController(self, didAuthenticateWithCode: code)
-        oauth2Service.fetchOAuthToken(withCode: code) { result in
+        UIBlockingProgressHUD.show()
+        self.delegate?.didAuthenticate(self)
```
Тут `self` необязательно использовать, поэтому лучше его убрать.
Подробнее можешь почитать в [код стайле](https://github.com/tutu-ru/swift-style-guide?tab=readme-ov-file#Комментарии)

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -43,55 +40,19 @@ final class WebViewViewController: UIViewController {
     // MARK: - UpdateProgress
     
     private func updateProgress() {
         progressView.progress = Float(webView.estimatedProgress)
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on ImageFeed/NetworkServices/OAuth2TokenStorage.swift:
```diff
@@ -5,23 +5,33 @@
+    static let shared = OAuth2TokenStorage()
+    private let keyChain = KeychainWrapper.standard
     
     var token: String? {
```
Не забывай организовывать код следующим образом:

- Публичные переменные
- IBOutlet
- Приватные переменные
- Публичные методы(сначала `init`ы, `override`ы, и потом остальные)
- Приватные методы
- IBAction

Подробнее почитать можно тут: [Организация кода](https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

helecta commented on ImageFeed/NetworkServices/OAuth2TokenStorage.swift:
```diff
@@ -5,23 +5,33 @@
+    }
     
-    private let tokenKey = "OAuth2Token"
+    static let shared = OAuth2TokenStorage()
```
Предполагается, что OAuth2TokenStorage будет использоваться в одном экземпляре, т.е. через shared. Тогда стоит добавить приватный init. Так мы реализуем паттерн  [singleton](https://www.hackingwithswift.com/example-code/language/what-is-a-singleton) .

helecta commented on ImageFeed/Profile/Models/UserResult.swift:
```diff
@@ -0,0 +1,22 @@
+struct UserResult: Codable {
+    let profileImage: ProfileImageSize
+    
+    enum CodingKeys: String, CodingKey {
```
Еще можно таким образом настроить JSONDecoder: https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  
Тогда CodingKeys будет не нужно

helecta commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -6,92 +6,183 @@
         let label = UILabel()
-        label.text = "Hello, World!"
         label.textColor = .ypWhite
         label.font = UIFont.systemFont(ofSize: 13, weight: .regular)
```
Это поле может быть длинным, в этом случае у тебя текст будет обрезаться.
Лучше тут сделать многострочный вариант строки. Для этого нужно сделать так:
```
label.numberOfLines = 0
```
В этом случае еще нужно добавить trailingAnchor для labelText

helecta commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -6,92 +6,183 @@
+    private func switchToSplashViewController() {
+        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
+              let window = windowScene.windows.first else {
+            fatalError("Invalid configuration of switchToSplashViewController")
```
Обычно в продакшене не используют fatalError, советую заменить на assertionFailure, тут подробнее:
 [https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions](https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions)

helecta commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -13,19 +13,24 @@ final class SplashViewController: UIViewController {
-    private let showAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
+    private let profileService = ProfileService.shared
+    private let profileImageService = ProfileImageService.shared
+    private let storage = OAuth2TokenStorage()
```
ты начал делать его синглтоном, поэтому тут его нужно использовать через .shared, и приватный init добавить как раз что бы не было таких ситуаций.

helecta commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -63,35 +75,53 @@ final class SplashViewController: UIViewController {
+        ProfileService.shared.fetchProfile(token) { [weak self] result in
+            DispatchQueue.main.async {
+                guard let self = self else { return }
+                UIBlockingProgressHUD.dismiss()
```
у тебя есть UIBlockingProgressHUD.dismiss(), но нет UIBlockingProgressHUD.show()
Если пользователь был авторизован и перезайдет в приложение, он не увидит лоудинг на момент загрузки профиля, т.к он нигде не начинается. Ты запускаешь его только на экране авторизации и вебвью. Но этих экранов тут не будет, если это повторный заход

EntityVadim committed:
- Добавил README

EntityVadim committed:
- Убрал self в строчке кода (self.delegate?.didAuthenticate(self))
- Организовывал код
- Добавил UIBlockingProgressHUD.show()
- OAuth2TokenStorage деделал синглетом
